### ISSUE TITLE
a way to tell that the virtualenv and Pipfile.lock are in sync?

### ISSUE THREAD

**Issue by cs01** [kennethreitz/pipenv#1080]
Ran the following

```bash
pipenv run ...
Creating a virtualenv for this project…
Using /usr/bin/python2.7 to create virtualenv…
⠋Running virtualenv with interpreter /usr/bin/python2.7
New python executable in /home/user/.virtualenvs/...-RVyaMy2x/bin/python2.7
Also creating executable in /home/user/.virtualenvs/...-RVyaMy2x/bin/python
Installing setuptools, pip, wheel...done.
...
ImportError: No module named yaml
```

##### Describe your environment

1. Ubuntu 16.04
1. Python 2.7.12
1. pipenv, version 8.3.2

##### Expected result

Expected one of the following:

1. Automatically install dependencies when `pipenv run ...` is called and all dependencies aren't in the virtualenv
1. Add optional flag to do 1
1. Add command to test if dependencies are installed so that I can run `pipenv install` before running `pipenv run ...` (as it stands now, `pipenv install` is fairly time consuming so I don't want to run it unconditionally) 
1. Modify `pipenv check` to check if all dependencies are installed and correct

I can try to make this change, just wanted to make sure it's something that you'd be okay merging.

##### Actual result

Virtualenv was installed but packages were not.

##### Steps to replicate

Run `pipenv run ...` on a python file with `Pipfile.lock` present, without running `pipenv install` first.

**Comment by techalchemy** [comment]
If you call something and it's not present then....what? Run becomes a blocking call until whatever that thing is and all of its dependencies are installed? Installation should be an intentional action, I would think...

Is there some use case I'm missing?  In the situation you described it sounds like you're saying we should be able to trust the `Pipfile.lock` -- so we can treat it like the user gave us the `--deploy` flag also? It seems like a lot of assumptions... Pipenv is really meant to manage a full environment of dependencies.  If you want to manage individual packages in separate, isolated environments i think `pipsi` is the tool you want.

**Comment by cs01** [comment]
My example use case is I want to run `file.py`, but `file.py` imports `yaml`, and I want to make sure when `file.py` is invoked that the `yaml` dependency is available and matches what's in `Pipfile.lock`. I believe this is the intended use of `pipenv`. Please correct me if I'm wrong.

I want to know *before* I run `file.py` whether its dependencies are installed. Is this unusual? And is there a way to do this?

As it stands now, `pipenv` will automatically create a virtual environment, then run the file, which of course won't work because `Pipenv.lock` dependencies weren't installed. 

> If you call something and it's not present then....what? Run becomes a blocking call until whatever that thing is and all of its dependencies are installed? 

If you call something and dependencies aren't available the following possibilities exist:

- The python script runs and fails at runtime (this is the current situation)
- pipenv realizes dependencies aren't compatible and aborts before invoking the script
- pipenv installs the necessary dependencies and invokes the script

So yes, it's a blocking call, but dependencies need to be installed at some point, so why not automate it if they aren't there already?

> Installation should be an intentional action, I would think...

In my original comment, I laid out 4 ways to fix the problem, 3 of which require the action to be intentional.

**Comment by techalchemy** [comment]
> My example use case is I want to run file.py, but file.py imports yaml, and I want to make sure when file.py is invoked that the yaml dependency is available and matches what's in Pipfile.lock. I believe this is the intended use of pipenv. Please correct me if I'm wrong.

This is not accurate.  Pipenv is not a code introspection tool, and does not run inside your virtual environment, so this would not be possible.  When pipenv creates your virtualenv, or when it installs a new thing into it, those are the times when it can do dependency resolution and installation.  Nothing about pipenv claims to resolve dependencies when you invoke them.

> As it stands now, pipenv will automatically create a virtual environment, then run the file, which of course won't work because Pipenv.lock dependencies weren't installed.

This is the only bug here.  Pipenv shouldn't create a virtualenv without installing your dependencies.  It should either not create the virtualenv and tell you to run _pipenv install_, or it should create the virtualenv and install everything in your pipfile.

> pipenv realizes the virtual environment's state does not match Pipfile.lock, prints and error message, and aborts before invoking the script

Pipenv doesn't check the lockfile against the virtualenv, so this will never occur.

**Comment by cs01** [comment]
The three bullet points I listed were possible outcomes. I said the first one is the way things work now ("this is the current situation"). The other two are hypothetical ways the tool could work.

> Pipenv doesn't check the lockfile against the virtualenv, so this will never occur.

I know. This was a possible outcome if the feature were added to pipenv. The third bullet also never occurs, but was a possible outcome.

(And FWIW I was not trying to be snarky or hostile in #1089. I am genuinely interested in understanding how this tool fits in with the other tools. 

[pip](https://github.com/pypa/pip) and [virtualenv](https://github.com/pypa/virtualenv) are both owned by the @pypa user, so it surprised me that an official tool like `pipenv` doesn't live there. I take issue because it concerns me that an officially recommended tool is controlled by an individual and not the community.  Now it's entirely possible Kenneth doesn't have the last say on what goes in this repo and what doesn't. But I don't see any documentation on how decisions and changes are made to `pipenv`, so I asked the question. 

I am only asking these questions because of the massive importance and influence a tool like `pipenv` has in the programming world. Hundreds of thousands/millions of developers rely or will rely on it.

`pypa/pip` claims it is "The PyPA recommended tool for installing Python packages" while pipenv says it's "the officially recommended Python packaging tool from Python.org". These are very similar, but not identical I guess. In any case it confused me. 

My interpretation of [pypa/Pipfile](https://github.com/pypa/pipfile) was that `Pipfile` is basically not ready for use because the specification is not entirely settled on. If that's not true, I am happy to be corrected. Again, I'm not trying to criticize, just trying to understand. I totally get that software is always in development, but that is different from deploying software based on a specification that warns you that it's unfinished. And if that's the case, that's fine, but maybe it should be noted that deploying Pipfile's to production is not recommended until the specification is settled on, etc.

In any case, it confused me and I wanted to get more info.  I would like to continue the conversation in that ticket if possible. Thanks for taking the time to respond, and thanks for maintaining pipenv.)

**Comment by techalchemy** [comment]
@cs01 thank you for clarifying, I am working on this issue, and if you do want to carry on the discussion in #1089 I will re-open it.  Sorry for the misunderstanding, we can certainly have that discussion

**Comment by cs01** [comment]
thanks, #1089 is opened but still locked

**Comment by cs01** [comment]
Can we either add to `pipenv check` or make a new command such as `pipenv isready` that does the following?

1. Captures the output of `pipenv run pip freeze` (if the virtualenv exists)
1. Captures the `[packages]` in `Pipfile`
1. Compares the two
1. Prints `Ready!` (and returns 0) if all packages are installed in the virtualenv. Prints `Needs installation. Run pipenv install.` (and returns nonzero)

My use case:
I have a complicated build, some of which involves python scripts, but don't want to have to run `pipenv install` over and over when it's not needed. It would be good if I could test if the install is needed.

If there's already a way to do this that I'm missing, please let me know.

**Comment by kennethreitz** [comment]
I like the idea of adding this to check. No new commands, though.

**Comment by erinxocon** [comment]
@kennethreitz `--validate` flag on the check command?

**Comment by kennethreitz** [comment]
Why have a flag? Just add a step to check

**Comment by cs01** [comment]
Which commit was this added in?

**Comment by kennethreitz** [comment]
not implemented yet, just noted as a potential feature to add

**Comment by erinxocon** [comment]
It has not been addressed yet.  I'm reopening until I fix it tonight.

**Comment by kennethreitz** [comment]
just because someone suggests something doesn't mean it deserves an open issue

**Comment by cs01** [comment]
How is the backlog of features on a project with a group of distributed developers, such as pipenv, managed unless they are somehow documented? And how is do devs know who is working on what, and what needs attention?

As the creator of the issue, my expectation is that either I close the ticket, or someone leaves a comment saying why it's being closed, or I see it's referenced in a closed pull request. I would not expect it to be closed if it's not finished. It looks like the expectations of how issues are managed is different between me, you and @erinxocon  A blurb on the CONTRIBUTING page would help to lay out expectations for everyone up front and make the issue reporting/closure process a little smoother.

Thanks again for making pipenv, it's got a great API. :+1:

**Comment by kennethreitz** [comment]
i don’t care what your expectation is, i’m closing it. thank you — it has been noted, as i said :)

**Comment by kennethreitz** [comment]
it would remain open if it was a “def will implement” thing, but it’s not.

**Comment by kennethreitz** [comment]
it’s a cool idea i’ll ponder

**Comment by nfisher** [comment]
It would be super useful for speeding up CI builds if `sync` had a flag to exit if everything is in sync.   A lot of SaaS CI platforms include code caches such as Travis, AWS codebuild, Bitbucket Pipelines, etc. The code cache is restored at the beginning of a run and if the dependencies didn't require updates it could shave 30s or more off of the build.

As it is `pipenv sync --dev` takes about 30s with a private build. It contains 100 or so deps in total. If `pipenv verify` (or whatever) had a return code of `1` on "out of sync" and `0` when it's "in sync".

Basically it would allow me to write something like this;

```
pipenv verify --dev
if [ $? -eq 1 ]; then
pipenv sync --dev
fi
```

If pipenv did the above internally that would be even better!

I will look at doing a diff between `pipenv lock -r` and `pip freeze` in the interim.
